workers/unity/Assets/Fps/Scripts/Config/MapBuilder.cs

```diff
@@ -12,53 +13,116 @@ public class MapBuilder : MonoBehaviour
 
     private const int TileSeparation = 36;
-    private const float GroundHeight = 0f;
-    private const float HeightOffset = 0.5f;
 
     private int groundWidth;
-    private const int WallHeight = 4;
+    private const int BlockSize = 4;
 
     private GameObject[] centreTiles;
     private GameObject[] levelTiles;
     private GameObject groundTile;
     private GameObject groundEdge;
     private GameObject surroundWall;
+    private GameObject cornerPiece;
 
     private Transform tileParentTransform;
     private Transform groundParentTransform;
     private Transform surroundParentTransform;
     private Transform cubeParentTransform;
+    private Transform spawnPointSystemTransform;
 
     private const string TileParentName = "TileParent";
     private const string GroundParentName = "GroundParent";
     private const string SurroundParentName = "SurroundParent";
-    private const string CubeParentName = "CubeParent";
+    private const string SpawnPointSystemName = "SpawnPointSystem";
 
     private const string LevelTilePath = "Prefabs/Level/Tiles";
     private const string GroundTilePath = "Prefabs/Level/Ground/Ground4x4";
     private const string GroundEdgePath = "Prefabs/Level/Ground/Edge";
     private const string SurroundPath = "Prefabs/Level/Surround/Wall";
+    private const string CornerPath = "Prefabs/Level/Surround/Corner";
 
     // Central tiles are hardcoded.
     private const string CentreTile0 = "Prefabs/Level/Tiles/Centre0";
     private const string CentreTile1 = "Prefabs/Level/Tiles/Centre1";
     private const string CentreTile2 = "Prefabs/Level/Tiles/Centre2";
     private const string CentreTile3 = "Prefabs/Level/Tiles/Centre3";
 
+    private int groundLayers => (Layers - 1) / 4 + 1;
+
 #if UNITY_EDITOR
-    public void Build()
+    public void CleanAndBuild()
     {
+        if (TryLoadResources() == false)
+        {
+            Debug.LogError("Generation aborted due to error");
+            return;
+        }
+
+        Clean();
+
+        InitializeGroupsAndComponents();
         Random.InitState(Seed.GetHashCode());
+
+        var originalPosition = transform.position;
+        var originalRotation = transform.rotation;
+        transform.position = Vector3.zero;
+        transform.rotation = Quaternion.identity;
         PlaceTiles();
         PlaceGround();
-        PlaceSurround();
+        FillSurround();
         MakeLevelObjectStatic();
-        Debug.Log($"Finished building world of size: {groundWidth} x {groundWidth}");
+
+        transform.position = originalPosition;
+        transform.rotation = originalRotation;
+
+        var worldSize = groundLayers * TileSeparation * 4 * 2 + 4;
+        Debug.Log($"Finished building world of size: {worldSize} x {worldSize}");
     }
 
-    private void PlaceTiles()
+    private void InitializeGroupsAndComponents()
+    {
+        if (!gameObject.GetComponent<TileSettings>())
+        {
+            gameObject.AddComponent<TileSettings>();
```

Comment:
Instead of this would it be better to use a `[RequireComponent(typeof(TileSettings))]` on this Monobehaviour? 

[RequireComponent docs](https://docs.unity3d.com/ScriptReference/RequireComponent.html)

Reply:
Probably would be, yes! It does mean however that anyone already using the MapBuilder component in their project will get a compile error.

>Note that RequireComponent only checks for missing dependencies during the moment the component is added to a GameObject. Existing instances of the component whose GameObject lacks the new dependencies will not have those dependencies automatically added.

Reply:
Actually so long as my draw distance branch goes in, the TileSettings object is no longer necessary as its function is shifted to a singleton scriptable object.